### PR TITLE
Fixes, improvements, and changes for Intel(R) Open Image Denoise

### PR DESCRIPTION
**Pull request by atafra**
This branch contains several fixes, improvements, and changes required for Intel(R) Open Image Denoise. Please see the individual commits for more details.

### CODE REVIEW

File: src/cpu/xbyak/xbyak.h [hunk 1]
```diff
@@ -242,7 +242,7 @@ class Error : public std::exception {
 	{
 		if (err_ < 0 || err_ > ERR_INTERNAL) {
 			fprintf(stderr, "bad err=%d in Xbyak::Error\n", err_);
-			exit(1);
```
Comment (rsdubtso) [inline src/cpu/xbyak/xbyak.h]: Please take this change to Xbyak if this change is absolutely necessary.

File: src/cpu/jit_avx512_common_convolution_winograd.cpp [hunk 2]
```diff
@@ -922,8 +922,8 @@ void _jit_avx512_common_convolution_winograd_t<is_fwd>::_execute_data_W_S_G_D(
             last_slice_bias[oc] = bias(jcp.dimM / jcp.dimM_simd_block - 1, oc);
     }
 
-    PRAGMA_OMP(parallel) {
-        parallel_nd_in_omp(jcp.mb, jcp.dimK_nb_block, jcp.dimK_block,
```
Comment (rsdubtso) [inline src/cpu/jit_avx512_common_convolution_winograd.cpp]: Do you have performance data that demonstrates changes in this commit do not negatively affect performance with OpenMP threading?

File: src/cpu/wino_reorder.hpp [hunk 3]
```diff
@@ -187,10 +187,10 @@ struct wino_reorder_t : public primitive_impl_t {
             out_data_t *__restrict _out
                     = tmp_wei + (iic * nb_oc_ + ob) * oc_block_;
 
-            parallel_nd(size_wspace_, [&](int i) { wspace[i] = 0.f; });
```
Comment (rsdubtso) [inline src/cpu/wino_reorder.hpp]: I believe the claim in the commit message, but how does this affect other cases? It also would be interesting to know how it affects different compilers.

File: src/common/CMakeLists.txt [hunk 4]
```diff
@@ -23,5 +23,8 @@ file(GLOB_RECURSE SOURCES
 
 set(OBJ_LIB ${LIB_NAME}_common)
 add_library(${OBJ_LIB} OBJECT ${SOURCES})
+set_target_properties(${OBJ_LIB} PROPERTIES COMPILE_FLAGS "${DNNL_ARCH_OPT_FLAGS}")
```
Comment (rsdubtso) [inline src/common/CMakeLists.txt]: Can you please explain why this change is needed?

File: CMakeLists.txt [hunk 5]
```diff
@@ -75,18 +75,18 @@ if(UNIX OR MINGW)
...
-include("cmake/options.cmake")
-include("cmake/OpenMP.cmake")
-include("cmake/TBB.cmake")
-include("cmake/OpenCL.cmake")
-include("cmake/platform.cmake")
-include("cmake/SDL.cmake")
-include("cmake/MKL.cmake")
-include("cmake/Doxygen.cmake")
-include("cmake/version.cmake")
+include(mkldnn_compat)
```
Comment (rsdubtso) [inline CMakeLists.txt]: Have you verified that this does not break when building DNNL as a subproject?

File: src/common/scratchpad.cpp [hunk 6]
```diff
@@ -74,14 +74,14 @@ struct global_scratchpad_t : public scratchpad_t {
...
+    static size_t size_;
+    static unsigned int reference_count_;
 };
 
-thread_local char *global_scratchpad_t::scratchpad_ = nullptr;
-thread_local size_t global_scratchpad_t::size_ = 0;
-thread_local unsigned int global_scratchpad_t::reference_count_ = 0;
+char *global_scratchpad_t::scratchpad_ = nullptr;
+size_t global_scratchpad_t::size_ = 0;
+unsigned int global_scratchpad_t::reference_count_ = 0;
```
Comment (dzarukin) [inline src/common/scratchpad.cpp]: This sounds to me as a severe library problem, which should have a separate test somewhere in `tests/other".

Comment (mgouicem) [inline src/common/scratchpad.cpp]: This behavior of the `global_scratchpad` in intended.
The global scratchpad is not thread safe, so creating a primitive using the global scratchpad in one thread and executing it in another fails (whether the buffer is `thread_local` or not).
The `thread_local` buffers enable some level of concurrency though (it works when you spawn multiple threads, and each thread creates and runs its own primitives). 
When creating and executing primitives in different threads, the user can either use concurrent scratchpad or pass the scratchpad as a parameter to the primitive execution.

In the example you mention (1 thread creates primitives, 2 threads execute them), it is true that you avoid a segfault, but the result will be incorrect (since two primitives are concurrently accessing the same buffer). For such a scenario, the global scratchpad should not be used.
However, if we remove the `thread_local`, it can cause segfaults in a currently supported scenario (2 threads, each creating and executing their own primitives).

Comment (dzarukin) [inline src/common/scratchpad.cpp]: This sounds to me as a major library problem, which should have a separate test somewhere in `tests/other".

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (atafra) [reply src/cpu/xbyak/xbyak.h]: Yes, this is absolutely necessary. Otherwise, MKL-DNN and other libraries that link it statically will be rejected from Fedora and other distributions. We're not using Xbyak directly, so please take care of taking this to Xbyak if you think it's necessary, but since MKL-DNN uses a copy of it, I don't see why this can't be merged immediately.

File: src/cpu/jit_avx512_common_convolution_winograd.cpp (hunk 2 above)
Comment (atafra) [reply src/cpu/jit_avx512_common_convolution_winograd.cpp]: No, we don't have performance data with OpenMP as we support only TBB. If you find any performance regressions, you could keep both versions using ifdefs. It would be a bit ugly, but it would avoid any potential issues.

File: src/cpu/wino_reorder.hpp (hunk 3 above)
Comment (atafra) [reply src/cpu/wino_reorder.hpp]: We don't have performance data for other workloads, but it spawns an excessive amount of very small tasks, which have a too high overhead. We tested this with ICC, but I don't expect significantly different results with other compilers. With parallel_nd we have an initialization time of more than 800 ms for a U-Net-like architecture, which is too much. A 4x speedup by avoiding parallel execution indicates that there's barely any parallelism available in these loops, and the overhead of parallel_nd is huge.

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (rsdubtso) [reply src/cpu/xbyak/xbyak.h]: We can handle talking to Herumi, but we are not merging this until we have a solution that works for Xbyak. We are keeping an exact copy of Xbyak to make merges easy. While I agree with the 'no-exits statement', this is a corner case and exit is never called in practice.

File: src/cpu/jit_avx512_common_convolution_winograd.cpp (hunk 2 above)
Comment (rsdubtso) [reply src/cpu/jit_avx512_common_convolution_winograd.cpp]: We cannot merge this then.

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (atafra) [reply src/common/CMakeLists.txt]: For Open Image Denoise we're directly using the compile parameters from MKL-DNN, but we can't always use ISAs that may not be available on all platforms, e.g. SSE4.1, so we need to compile a portion of the library without these extra ISAs. Having the possibility to apply these ISA options to only certain source files enables us to detect ISA support in the library and report an error instead of crashing due to executing unsupported instructions. This change probably would not have direct benefits for MKL-DNN right now, but it would make adding this behavior straightforward in the future. Compiling the entire MKL-DNN library for a specific ISA is not flexible enough both performance and backward compatibility wise.

File: CMakeLists.txt (hunk 5 above)
Comment (atafra) [reply CMakeLists.txt]: We're already building DNNL as a subproject and so far we haven't seen any issues.

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (atafra) [reply src/common/scratchpad.cpp]: I think you misunderstood the issue. You don't need concurrent execution at all to cause a crash with the current behavior. There's no concurrency, no multiple threads executing at the same time. The crash happens if you initialize a primitive on thread A, and later invoke that primitive from thread B, but thread A and thread B are *not* concurrent, there's no issue of thread safety, everything is synchronized. Removing thread_local does *not* cause a crash in this scenario, and this is how Open Image Denoise is currently working. I think this is quite dangerous behavior and took quite some time to figure out why a segfault is happening. If multiple threads want to execute primitives, wouldn't a concurrent scratchpad be the right option instead?

File: src/cpu/jit_avx512_common_convolution_winograd.cpp (hunk 2 above)
Comment (atafra) [reply src/cpu/jit_avx512_common_convolution_winograd.cpp]: Is this because you've found performance regressions with OpenMP?

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (atafra) [reply src/cpu/xbyak/xbyak.h]: Thanks for taking this issue upstream, but I disagree that this is corner case. If the library links exit, than in practice it *can* be called, and this disqualifies it automatically from Fedora and potentially other distributions because it fails with their testing tools. Open Image Denoise was rejected as well because of this issue in MKL-DNN.

File: src/cpu/jit_avx512_common_convolution_winograd.cpp (hunk 2 above)
Comment (rsdubtso) [reply src/cpu/jit_avx512_common_convolution_winograd.cpp]: No, because we don't know if this change introduces regressions or not. So until we have performance data and are sure that there are no regressions, we cannot merge this.

File: CMakeLists.txt (hunk 5 above)
Comment (vpirogov) [reply CMakeLists.txt]: The changes in CMakes seem to break building the library as a project though.

```
$ cmake --version
cmake version 2.8.11
$ cmake ..
-- CMAKE_BUILD_TYPE is unset, defaulting to Release
-- GPU support is disabled
CMake Error at cmake/utils.cmake:99 (target_link_libraries):
  Object library target "dnnl_common" may not link to anything.
Call Stack (most recent call first):
  src/common/CMakeLists.txt:27 (target_link_libraries_build)
 
 
-- Configuring incomplete, errors occurred!
```

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (rsdubtso) [reply src/cpu/xbyak/xbyak.h]: I understand Fedora's stance, and I am sorry that that Open Image Denoise was rejected, but this  does not sound like a reason for this change to go around the rules we have for interacting with Xbyak upstream. I've opened the issue here: https://github.com/herumi/xbyak/issues/83

File: CMakeLists.txt (hunk 5 above)
Comment (atafra) [reply CMakeLists.txt]: It builds fine on my machine as a project as well, but I'm using a much more recent version of CMake (3.13.4).

File: src/cpu/jit_avx512_common_convolution_winograd.cpp (hunk 2 above)
Comment (atafra) [reply src/cpu/jit_avx512_common_convolution_winograd.cpp]: Sure, that makes sense. But unfortunately we don't have resources to provide our own performance data for OMP or other workloads. Could you please check this with your benchmark suite?

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (mgouicem) [reply src/common/scratchpad.cpp]: Out of curiosity, why do you create the primitives and execute them in separate threads if there is no concurrency?

Comment (atafra) [reply src/common/scratchpad.cpp]: It's not Open Image Denoise what directly invokes this from different threads, but other libraries and applications that use Open Image Denoise. E.g. OSPRay (http://www.ospray.org/), which executes denoising asynchronously with other rendering operations on a thread different from the main thread. So actually there is concurrency, but there is always only one thread invoking MKL-DNN operations at a time. The other threads are executing other tasks.

File: cmake/platform.cmake [hunk 7]
```diff
@@ -143,13 +155,20 @@ elseif(UNIX OR MINGW)
         append(CMAKE_CCXX_NOWARN_FLAGS "-diag-disable:15335")
         # disable: foo has been targeted for automatic cpu dispatch
         append(CMAKE_CCXX_NOWARN_FLAGS "-diag-disable:15009")
+        # disable optimizations in debug mode
+        append(CMAKE_CXX_FLAGS_DEBUG "-O0")
     endif()
+    # disable assertions
+    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -DNDEBUG")
```
Comment (echeresh) [inline cmake/platform.cmake]: CMake passes `NDEBUG` in the release configuration by default. Are there cases when we have to handle this manually?

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (echeresh) [reply src/common/CMakeLists.txt]: Could we handle this by setting `CMAKE_CXX_FLAGS` in `src/CMakeLists.txt`? The issue with the current solution is that we have to duplicate `set_target_properties` for every sub-directory of `src`: `src/common`, `src/cpu`, `src/ocl` and any other directory we add in future.

On the other hand, if the goal is to limit `DNNL_ARCH_OPT_FLAGS` to the `src` directory we could just move `append(CMAKE_CXX_FLAGS "${DNNL_ARCH_OPT_FLAGS}")` from `cmake/platform.cmake` to `src/CMakeLists.txt`.

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (densamoilov) [reply src/common/scratchpad.cpp]: Currently, if scratchpad mode is "library" then a user has to create and execute a primitive in the same thread otherwise it can crash if the primitive uses global scratchpad. 
This change also makes primitive creation not thread-safe.

File: src/common/CMakeLists.txt [hunk 8]
```diff
@@ -23,5 +23,8 @@ file(GLOB_RECURSE SOURCES
...
+target_link_libraries_build(${OBJ_LIB}
+    "${EXTRA_SHARED_LIBS};${EXTRA_STATIC_LIBS}")
```
Comment (echeresh) [inline src/common/CMakeLists.txt]: As I know `target_link_libraries` didn't work with object libraries before CMake 3.12 (DNNL supports 2.8.11+).

Can you please explain what is the goal of this change? Now (with 2.8.11) we can't propagate dependencies from object libraries and have to handle that globally in `src/CMakeLists.txt`. Do you see any issues with that?

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (atafra) [reply src/common/scratchpad.cpp]: Yes, but it shouldn't crash if there's no concurrency.

Comment (rsdubtso) [reply src/common/scratchpad.cpp]: I would suggest for OID to switch to user scratchpads. This was not possible in 0.x, but no there's no reason not to if the usage model is different from what the library supports.

Comment (densamoilov) [reply src/common/scratchpad.cpp]: Agree, it shouldn't crash with this change, but it fixes one thing and breaks another. I agree with @rsdubtso on his suggestion to use user scratchpad mode.

Comment (atafra) [reply src/common/scratchpad.cpp]: We've already switched to user scratchpads, but I still think the current behavior of MKL-DNN is very confusing and error-prone. Is this documented somewhere?

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (atafra) [reply src/common/CMakeLists.txt]: Yes, moving `append(CMAKE_CXX_FLAGS "${DNNL_ARCH_OPT_FLAGS}")` to `src/CMakeLists.txt` should work too.

File: CMakeLists.txt [hunk 9]
```diff
@@ -75,18 +75,18 @@ if(UNIX OR MINGW)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
 endif()
 
-include("cmake/mkldnn_compat.cmake")
```
[the remark this answers is not in the document]
Comment (atafra) [reply CMakeLists.txt]: If full paths are used (`cmake/` prefix), including `.cmake` files works only within MKL-DNN, we can't include them from Open Image Denoise (which has MKL-DNN as a subproject) because CMake includes are unfortunately relative to the source directory, not the file that contains the include. Dropping full paths fixes this issue.

File: cmake/SDL.cmake [hunk 10]
```diff
@@ -60,3 +58,6 @@ if(UNIX)
 elseif(MSVC AND ${CMAKE_CXX_COMPILER_ID} STREQUAL MSVC)
     set(CMAKE_CCXX_FLAGS "/guard:cf")
 endif()
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${CMAKE_CCXX_FLAGS}")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${CMAKE_CCXX_FLAGS}")
```
[the remark this answers is not in the document]
Comment (atafra) [reply cmake/SDL.cmake]: Sure, `append` could be used instead.

File: src/common/CMakeLists.txt (hunk 8 above)
Comment (atafra) [reply src/common/CMakeLists.txt]: I didn't know that this is such a recent CMake feature.

We have a modified `TBB.cmake`, which sets the TBB include directories for the TBB targets, not globally, so we need to propagate these to the MKL-DNN libraries through `target_link_libraries_build`. With older MKL-DNN versions this worked fine because there was a single static/dynamic MKL-DNN library, and only one `target_link_libraries_build` was needed. Now this is more difficult due to the multiple object libraries. Is there any other way to propagate the TBB target properties to the MKL-DNN object libraries?

File: cmake/platform.cmake (hunk 7 above)
Comment (atafra) [reply cmake/platform.cmake]: At some point in the past it was needed for some reason, but not anymore. So this can be removed.

File: src/common/CMakeLists.txt (hunk 8 above)
Comment (echeresh) [reply src/common/CMakeLists.txt]: We use global `include_directories` to propagate include directories from `TBB.cmake` to all source files:
https://github.com/intel/mkl-dnn/blob/master/cmake/TBB.cmake#L38

Doesn't it work for you? If you need to limit the scope you can call `include_directories` from a specific sub-directory.

Comment (atafra) [reply src/common/CMakeLists.txt]: This doesn't work for us because it's not `include_directories` that we need to set but `INTERFACE_INCLUDE_DIRECTORIES`.

File: CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (echeresh) [reply CMakeLists.txt]: Thanks, I see. But we have a sub-project "test" in the repository:
https://github.com/intel/mkl-dnn/blob/master/tests/other/subproject/CMakeLists.txt

and it works fine even without this change. I'm not sure what might be the reason that we are not seeing the issues you described with our small sub-project...

[the remark this answers is not in the document]
Comment (rsdubtso) [reply CMakeLists.txt]: This is because OIDn [includes files from DNNL directly](https://github.com/OpenImageDenoise/oidn/blob/dabfd9c80101edae9d25a710160d12d6d963c591/CMakeLists.txt#L90-L94)

[the remark this answers is not in the document]
Comment (atafra) [reply CMakeLists.txt]: What you describe is completely different from our case. MKL-DNN is the subproject of OIDN, not the other way around. But the issue is not related to subprojects. In OIDN we directly include some `.cmake` files from MKL-DNN, which fails if that file includes other files with full path.

**Comment by atafra** [comment]
I see that most of the discussion is about the CMake changes. Please note that those are by far the least important ones.

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (vpirogov) [reply src/common/scratchpad.cpp]: Scratchpad has it's own section in the [developer guide](http://intel.github.io/mkl-dnn/dev_guide_attributes_scratchpad.html), however the scenario you have is not covered there. I agree that this behavior should be documented.

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (vpirogov) [reply src/cpu/xbyak/xbyak.h]: The changes are implemented upstream. We will pull this in shortly.

**Comment by atafra** [comment]
I've reverted all CMake changes as we've decided to use our own custom CMake files to build MKL-DNN for Open Image Denoise, so these changes are not required anymore.

**Comment by atafra** [comment]
Here's the DNNL log for our network:
[dnnl_log.txt](https://github.com/intel/mkl-dnn/files/3768687/dnnl_log.txt)

File: src/common/CMakeLists.txt (hunk 8 above)
Comment (vpirogov) [reply src/common/CMakeLists.txt]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-541941800)

File: cmake/platform.cmake (hunk 7 above)
Comment (vpirogov) [reply cmake/platform.cmake]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-544695471)

File: cmake/SDL.cmake (hunk 10 above)
[the remark this answers is not in the document]
Comment (vpirogov) [reply cmake/SDL.cmake]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-544695471)

File: CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (vpirogov) [reply CMakeLists.txt]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-544695471)

File: CMakeLists.txt (hunk 5 above)
Comment (vpirogov) [reply CMakeLists.txt]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-544695471)

File: src/common/CMakeLists.txt (hunk 4 above)
Comment (vpirogov) [reply src/common/CMakeLists.txt]: Closing the discussion as the changes [are not relevant anymore](https://github.com/intel/mkl-dnn/pull/572#issuecomment-544695471)

File: src/common/memory.cpp [hunk 11]
```diff
@@ -67,8 +67,7 @@ dnnl_memory::dnnl_memory(dnnl::impl::engine_t *engine,
     status_t status = engine->create_memory_storage(
             &memory_storage_ptr, flags, size, handle);
 
-    assert(status == status::success);
-    MAYBE_UNUSED(status);
+    if (status != status::success) return;
```
Comment (vpirogov) [inline src/common/memory.cpp]: Just a random place to put a comment for commit a336de93a2006586f524ab122e5e48923f14f141.

@rsdubtso is massaging the code to get it promoted.

File: src/cpu/xbyak/xbyak.h (hunk 1 above)
Comment (vpirogov) [reply src/cpu/xbyak/xbyak.h]: The change is promoted in f32e21b48948d4bc5402b77e2f6bd99f2eaa76ea.

File: src/common/scratchpad.cpp (hunk 6 above)
Comment (vpirogov) [reply src/common/scratchpad.cpp]: Scratchpad section in [developer guide](http://intel.github.io/mkl-dnn/dev_guide_attributes_scratchpad.html) is extended with additional information on thread safety.

File: src/common/memory.cpp (hunk 11 above)
Comment (vpirogov) [reply src/common/memory.cpp]: The patch is [promoted](9ab68b1b33319f6d996a6f156ae9c7954557a14e)

**Comment by vpirogov** [comment]
Here's the disposition of the changes in this PR based on the discussions we had so far.

| Commit | Disposition
|  :-           |  :-
| e18623ef174ae2bfa349ed3fbd2eb3438bda9a5e | Pending performance evaluation.
| bef6dc738f2cb16bb04249dbf7e6b1fec30b1aa6 | Pending performance evaluation.
| bae345e836c962a35dd0c49c5ad467673fe9fb5c | Will not promote. OIDN should switch to user-provided scratchpads. Documentation updated to clarify thread-safety considerations when using library-managed scratchpads.
| 08ef5325a58c2ed3bd334f76c444195857556eab | Implemented upstream and promoted
| a336de93a2006586f524ab122e5e48923f14f141 | Promoted
| 332ce0fa25650aa1103432ce9adaa480c80c8b97 | Will not promote. OIDN to use custom CMakefile

**Comment by vpirogov** [comment]
TBB patches were promoted. With that all the changes are dispositioned and I'm closing the PR.

Commit | Disposition
-- | --
e18623e | Promoted.
bef6dc7 | Promoted.
bae345e | Will not promote. OIDN should switch to user-provided scratchpads. Documentation updated to clarify thread-safety considerations when using library-managed scratchpads.
08ef532 | Implemented upstream and promoted
a336de9 | Promoted
332ce0f | Will not promote. OIDN to use custom CMakefile

Thank you for your contribution.
